Use redis recursive scan with match instead of keys to reduce pressure on redis

whatl3y opened the pull request:
Resolves #263 

I didn't write any new tests since based on a quick review the current test coverage should catch any issues that exist with the new `connection.getKeys`. I also updated an existing test to use the new method as well. All places in the repo using `connection.redis.keys` should now use `connection.getKeys` which uses recursive scan instead of keys.

evantahler commented:
Fast work! 

I agree that the existing tests act as enough of an "integration test" for the new method, but I would still like to see a "unit" test for `connection.getKeys`... something that shows that only keys that match the matcher are returned, and that the "loop" works, that you will get more than 1 loop of the cursor.

whatl3y commented:
@evantahler fair enough, added unit test in https://github.com/taskrabbit/node-resque/pull/264/commits/10c304e9ca2cdd0fee71a8b3982035549961fd6a

You mention adding a test to confirm the loop works. I'm assuming the traditional way to do this would be to check that the cursor is updated/incremented appropriately until it gets back to it's terminating value (`0` in our case). Since the cursor returned from `scan` has no guarantees of what will be returned and should be treated as random (see https://redis.io/commands/scan#scan-basic-usage), I'm unsure how to effectively do that check. It also makes it difficult since I added the cursor to be internal to the function and not returned in any way by the function.

Will the current test I wrote suffice, or would you like an additional test?

evantahler commented:
Dang, those are some harsh limitations! Given all of that, I think your test is as good as can be done.

I'm now curious to how ioredis handles possible duplicates... but that's out of scope for this PR :D

evantahler commented:
Thanks!

evantahler commented:
released in https://github.com/taskrabbit/node-resque/releases/tag/v5.5.0
